src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,760 @@
+    iov = slices.map { |slice| LibC::IOVec.new(slice.to_unsafe, slice.size.to_u64) }
+    submit_and_wait :writev, timeout: timeout do |sqe|
+      sqe.value.fd = fd
+      sqe.value.addr = pointerof(iov).address
+      sqe.value.len = slices.size.to_u
+      sqe.value.off = offset
+    end
+  end
+
+  # Writes a slice of data to fd at the position `offset`, by default the current position.
+  def write(fd : Int32, slice : Bytes, offset : UInt64 = -1.to_u64!, *, timeout : ::Time::Span? = nil)
+    if IoUring.probe.operations.write?
+      submit_and_wait :write, timeout: timeout do |sqe|
+        sqe.value.fd = fd
+        sqe.value.addr = slice.to_unsafe.address
+        sqe.value.len = slice.size.to_u
+        sqe.value.off = offset
+      end
+    else
+      writev(fd, {slice}, offset, timeout: timeout)
+    end
+  end
+
+  # Reads a sequence o slices to fd. Each slice is read sequencially starting at `offset` of the file, by default the current position.
+  # The entire read is atomic in the sense that it is garanteed that all buffers will be read from the file one after the other
+  # even if other process is trying to read to the same file at the same time.
+  def readv(fd : Int32, slices : Array(Bytes), offset : UInt64 = -1.to_u64!, *, timeout : ::Time::Span? = nil)
+    iov = slices.map { |slice| LibC::IOVec.new(slice.to_unsafe, slice.size.to_u64) }
+    submit_and_wait :readv, timeout: timeout do |sqe|
+      sqe.value.fd = fd
+      sqe.value.addr = iov.to_unsafe.address
+      sqe.value.len = slices.size.to_u
+      sqe.value.off = offset
+    end
+  end
+
+  # :ditto:
+  def readv(fd : Int32, slices : Tuple(Bytes), offset : UInt64 = -1.to_u64!, *, timeout : ::Time::Span? = nil)
+    iov = slices.map { |slice| LibC::IOVec.new(slice.to_unsafe, slice.size.to_u64) }
+    submit_and_wait :readv, timeout: timeout do |sqe|
+      sqe.value.fd = fd
+      sqe.value.addr = pointerof(iov).address
+      sqe.value.len = slices.size.to_u
+      sqe.value.off = offset
+    end
+  end
+
+  # Read a slice of data to fd from the position `offset`, by default the current position.
+  def read(fd : Int32, slice : Bytes, offset : UInt64 = -1.to_u64!, *, timeout : ::Time::Span? = nil)
+    if IoUring.probe.operations.read?
+      submit_and_wait :read, timeout: timeout do |sqe|
+        sqe.value.fd = fd
+        sqe.value.addr = slice.to_unsafe.address
+        sqe.value.len = slice.size.to_u
+        sqe.value.off = offset
+      end
+    else
+      readv(fd, {slice}, offset, timeout: timeout)
+    end
+  end
+
+  def send(fd : Int32, slice : Bytes, *, timeout : ::Time::Span? = nil)
+    unless IoUring.probe.operations.send?
+      return write(fd, slice, timeout: timeout)
+    end
+
+    submit_and_wait :send, timeout: timeout do |sqe|
+      sqe.value.fd = fd
+      sqe.value.addr = slice.to_unsafe.address
+      sqe.value.len = slice.size.to_u
+    end
+  end
+
+  def recv(fd : Int32, slice : Bytes, *, timeout : ::Time::Span? = nil)
```

Comment:
`recvfrom` seems to be missing an implementation? 

Unfortunately it lacks an op in uring so an implementation needs to fall back to the more clumsy but powerful `recvmsg`

Reply:
I was saving those for a future PR when they are going to be used for the stdlib. For now only Nop, Timeout, CancelTimoue and PollAdd are in use.

Reply:
Ah, right. I suppose it is easier to get something mergeable this way even if it probably won't be optimal at this point. 

Some things probably won't be approaching optimal in any case until CI is upgraded to something more modern than 5.4, I suppose.